cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -268,18 +268,35 @@ protected void initGoalState(ClusterModel clusterModel, OptimizationOptions opti
     _isLowUtilization = avgUtilizationPercentage <= _balancingConstraint.lowUtilizationThreshold(resource());
     if (_isLowUtilization) {
       // Identify a typical broker capacity to be used in recommendations in case the cluster is over-provisioned.
-      int typicalBrokerId = _brokersAllowedReplicaMove.iterator().next();
+      // To make sure that the recommended brokers to drop in an heterogeneous cluster is a positive number, the typical broker must have
+      // the max capacity.
+      int typicalBrokerId = brokerIdWithMaxCapacity(clusterModel);
       double typicalCapacity = clusterModel.broker(typicalBrokerId).capacityFor(resource());
 
       // Any capacity greater than the allowed capacity may yield over-provisioning.
       double allowedCapacity = resourceUtilization / _balancingConstraint.lowUtilizationThreshold(resource());
       int allowedNumBrokers = (int) (allowedCapacity / typicalCapacity);
-      int numBrokersToDrop = _brokersAllowedReplicaMove.size() - allowedNumBrokers;
+      int numBrokersToDrop = Math.max(_brokersAllowedReplicaMove.size() - allowedNumBrokers, 1);
```

Comment:
@efeg 

What if `_brokersAllowedReplicaMove.size()`  equals `allowedNumBrokers`? Wouldn't this suggest removing a broker even when cluster size doesn't need any change?

Wouldn't it be safer to make this line look like this? 

```java
int numBrokersToDrop = Math.max(_brokersAllowedReplicaMove.size() - allowedNumBrokers, 0);
```